pkg/agent/util/ipset/ipset_test.go

```diff
@@ -0,0 +1,191 @@
+			}
+			fakeExec := &exectesting.FakeExec{
+				CommandScript: []exectesting.FakeCommandAction{
+					func(cmd string, args ...string) exec.Cmd { return exectesting.InitFakeCmd(&fcmd, cmd, args...) },
+				},
+			}
+			c := &Client{exec: fakeExec}
+			err := c.CreateIPSet(testIPSetName, tt.setType, tt.isIPv6)
+			assert.Equal(t, tt.expectedArgs, fcmd.Argv)
+			if tt.expectedErrOutput != "" {
+				assert.EqualError(t, err, tt.expectedErrOutput)
+			} else {
+				assert.NoError(t, err)
+			}
+		})
+	}
+}
+
+func TestDestroyIPSet(t *testing.T) {
+	tests := []struct {
+		name              string
+		err               error
+		output            string
+		expectedErrOutput string
+	}{
+		{
+			name: "Destroy set successfully",
+		},
+		{
+			name: "Destroy non-existent set, no error",
+			err:  errors.New("The set with the given name does not exist"),
+		},
+		{
+			name:              "Destroy set with other error",
+			err:               errors.New("some errors"),
+			output:            "error output",
+			expectedErrOutput: fmt.Sprintf("error destroying ipset %s, err: some errors, output: error output", testIPSetName),
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			fcmd := exectesting.FakeCmd{
+				CombinedOutputScript: []exectesting.FakeAction{func() ([]byte, []byte, error) { return []byte(tt.output), nil, tt.err }},
+			}
+			fakeExec := &exectesting.FakeExec{
+				CommandScript: []exectesting.FakeCommandAction{
+					func(cmd string, args ...string) exec.Cmd { return exectesting.InitFakeCmd(&fcmd, cmd, args...) },
+				},
+			}
+			c := &Client{exec: fakeExec}
+			err := c.DestroyIPSet(testIPSetName)
+			assert.Equal(t, []string{"ipset", "destroy", testIPSetName}, fcmd.Argv)
+			if tt.expectedErrOutput != "" {
+				assert.EqualError(t, err, tt.expectedErrOutput)
+			} else {
+				assert.NoError(t, err)
+			}
+		})
+	}
+}
+
+func TestAddEntry(t *testing.T) {
+	fcmd := exectesting.FakeCmd{
+		CombinedOutputScript: []exectesting.FakeAction{
+			func() ([]byte, []byte, error) { return []byte("error output"), nil, errors.New("some errors") },
+			func() ([]byte, []byte, error) { return nil, nil, nil },
+		},
+	}
+	fakeExec := &exectesting.FakeExec{
+		CommandScript: []exectesting.FakeCommandAction{
+			func(cmd string, args ...string) exec.Cmd { return exectesting.InitFakeCmd(&fcmd, cmd, args...) },
+			func(cmd string, args ...string) exec.Cmd { return exectesting.InitFakeCmd(&fcmd, cmd, args...) },
+		},
+	}
+	c := &Client{exec: fakeExec}
+	err := c.AddEntry(testIPSetName, "1..2.3.4")
+	assert.Equal(t, []string{"ipset", "add", testIPSetName, "1..2.3.4", "-exist"}, fcmd.Argv)
+	assert.EqualError(t, err, fmt.Sprintf("error adding entry 1..2.3.4 to ipset %s, err: some errors, output: error output", testIPSetName))
+	err = c.AddEntry(testIPSetName, "1.2.3.4")
+	assert.Equal(t, []string{"ipset", "add", testIPSetName, "1.2.3.4", "-exist"}, fcmd.Argv)
+	assert.NoError(t, err)
```

Comment:
why are we accessing `fcmd.Argv` directly instead of using the package the way it is supposed to be used? It may be the same thing here, but `fcmd.Argv` is populated at the time the fake command is initialized, while `CombinedOutputCalls` / `CombinedOutputLog` are populated as the command is executed.

```go
	c := &Client{exec: fakeExec}
	err := c.AddEntry(testIPSetName, "1..2.3.4")
	assert.Equal(t, 1, fcmd.CombinedOutputCalls)
	assert.Equal(t, []string{"ipset", "add", testIPSetName, "1..2.3.4", "-exist"}, fcmd.CombinedOutputLogs[0])
	assert.EqualError(t, err, fmt.Sprintf("error adding entry 1..2.3.4 to ipset %s, err: some errors, output: error output", testIPSetName))
	err = c.AddEntry(testIPSetName, "1.2.3.4")
	assert.Equal(t, 2, fcmd.CombinedOutputCalls)
	assert.Equal(t, []string{"ipset", "add", testIPSetName, "1.2.3.4", "-exist"}, fcmd.CombinedOutputLogs[1])
	assert.NoError(t, err)
```

Reply:
Make sense. Verifying the expectation from CombinedOutputLogs is better!